### DISCUSSION TITLE
interactive mode for PipelineRun

### DISCUSSION THREAD

**Discussion by QingJuBaiTang** [tektoncd/pipeline#4134] [Ideas]
The execution of tekton pipeline now is totally automated. That what tasks a pipeline includes, the order of the execution of tasks, and whether a taskrun should be retried after failed, and retry times, and so on is pre-defined in the yaml file.

In most cases, especially in CI cases, this pattern can satisfy demand very well, but in some cases, such as CD cases , the process needs human intervention, which can not be well supported now.

For example, we now have an deploy pipeline, prepare to deliver a new version to a host. Before that, we have passed some tasks (merge branch, code quality check, build and test), but finally failed in deploy task, Perhaps the reason of that is a lack of java environment in host, so the pipeline abort and failed. Thus user would have to check and fix the problem, and rerun the whole process, which is not friendly to user. A more reasonable way is to trigger to continue after user check and fix the problem.

so, I propose to introduce **"interactive mode"**, which can be declared in PipelineRun. With this mode on, it means PipelineRun would not be finished automatically until user confirm to close or cancel. 

In addition, these features below are supported in this mode:
1. under interactive mode, failed tasks should not caused the failure of pipeline, and the failed tasks can be retried by user in the future, after tasks execute success, pipeline recovered and be continued
2. under interactive mode, pipeline flow can be paused after some tasks are executed successfully, and then trigger to continue by user manually in the future.
3. under interactive mode, the task which executed successfully could be retried (this feature can be enabled by field, so as to have a good control of which tasks will be retried and to make the retry safety), unless the whole pipeline is closed.

**Comment by ScrapCodes** [comment]
Thank you for the proposal !

May be you can take a look at this proposal, https://github.com/tektoncd/community/pull/484 . It lets you partially execute a pipeline and also resume/retry a previously run pipeline.

**Reply by QingJuBaiTang** [reply]
Thanks for your reply.
partial execution is great, and I have two questions:
1. it cannot cover the 2nd case, which is a pre-defined pause. Is there any solution for it?
2. The another question is that  if we retry too many times, partial execution shouled create many prs correspondingly, and then one complete execution will be divided into many pieces, which is not intuitive for user to observe the whole status of execution.

**Reply by ScrapCodes** [reply]
Good points, 

There is another discussion on why we cannot update a completed pr : https://github.com/tektoncd/community/pull/422

**Reply by QingJuBaiTang** [reply]
Oh, I see. so the final solution to retry failed task is partial execution? is there a schedule about the release of this feature? and any ideas about pre-defined pause?

**Reply by ScrapCodes** [reply]
Can you explain the use cases for  pre-defined pause?

**Reply by QingJuBaiTang** [reply]
Ok, one scenario is canary release. We deploy a new version only into one host in first batch, and then we pause the process to observe if everything is ok. we continue the process, otherwise we stop and rollback. 
so there is a pause wait for user confirm to continue at first batch, I called it pre-defined pause. how to deal with problems like this in tekton?

**Comment by afrittoli** [comment]
Hello @QingJuBaiTang - thank you for sharing your idea and use case.

The need for human approval - or interaction with an external system in general - has come up before, see the following links:
- https://github.com/tektoncd/pipeline/issues/4074
- https://github.com/tektoncd/pipeline/issues/1136

Another way you can interact with external systems is through events. Tekton supports sending `CloudEvents` throughout the lifecycle of `TaskRuns` and `PipelineRuns`. Tekton also supports running `TaskRuns` and `PipelineRuns` as response to a trigger, through the `tektoncd/triggers` project.

Using events you can create distributed workflow, which may involve other systems. The downside of this approach is that your workflow does not fit a single Tekton pipeline anymore.

This PoC by the "Events Special Interest Group" might also be of interest for you: https://github.com/cdfoundation/sig-events/tree/main/poc - there we used events to connect Tekton and Keptn. Keptn provides there a manual approval step, which is used to trigger a new Tekton pipeline.

Finally, even if the use case is different, the work started at https://github.com/tektoncd/community/blob/main/teps/0042-taskrun-breakpoint-on-failure.md may provide some initial background for interactive execution, since it introduces the ability to interactively execute steps in a Task.  Doing the same for a pipeline might be more complex, since pipelines do not necessarily have a linear execution, but it's something worth discussing.

**Reply by QingJuBaiTang** [reply]
Thanks for your help, I will take some time to take a look at what you mentioned.
